File: ios-im-ui/Sources/FlareIMUI/Components/MessageContentView.swift
import SwiftUI

/// Context passed to every content renderer.
public struct FlareContentContext {
    public let isSelf: Bool
    public let previewMode: Bool
    public let senderName: String?
    public let mediaState: FlareMediaDownloadState?
    public let onMediaAction: ((FlareMessageContent) -> Void)?

    public init(
        isSelf: Bool,
        previewMode: Bool = false,
        senderName: String? = nil,
        mediaState: FlareMediaDownloadState? = nil,
        onMediaAction: ((FlareMessageContent) -> Void)? = nil
    ) {
        self.isSelf = isSelf
        self.previewMode = previewMode
        self.senderName = senderName
        self.mediaState = mediaState
        self.onMediaAction = onMediaAction
    }
}

public typealias FlareContentBuilder = (FlareMessageContent, FlareContentContext) -> AnyView

/// Registry for product content types (`vote`, `task`…). Built-in types are
/// rendered directly by ``MessageContentView``; register a builder to add or
/// override a type.
public enum FlareContentRegistry {
    private static var builders: [String: FlareContentBuilder] = [:]
    public static func register(_ type: String, _ builder: @escaping FlareContentBuilder) {
        builders[type] = builder
    }
    public static func unregister(_ type: String) { builders.removeValue(forKey: type) }
    public static func lookup(_ type: String) -> FlareContentBuilder? { builders[type] }
}

/// Content-type dispatcher — renders a message body by type. Spec:
/// Message/MessageContentView (`MessageContentView`).
public struct MessageContentView: View {
    private let content: FlareMessageContent
    private let ctx: FlareContentContext
    @Environment(\.colorScheme) private var scheme

    public init(
        content: FlareMessageContent,
        isSelf: Bool = false,
        previewMode: Bool = false,
        senderName: String? = nil,
        mediaState: FlareMediaDownloadState? = nil,
        onMediaAction: ((FlareMessageContent) -> Void)? = nil
    ) {
        self.content = content
        self.ctx = FlareContentContext(
            isSelf: isSelf, previewMode: previewMode, senderName: senderName,
            mediaState: mediaState, onMediaAction: onMediaAction)
    }

    public var body: some View {
        let colors = FlareColors.of(scheme)
        let onBubble = ctx.isSelf ? Color.white : colors.textPrimary

        if let custom = FlareContentRegistry.lookup(content.type) {
            custom(content, ctx)
        } else {
            switch content {
            case let c as FlareTextContent:
                Text(c.text)
                    .font(.system(size: FlareSizes.fontSizeXl))
                    .lineSpacing(4)
                    .foregroundColor(onBubble)
            case let c as FlareEmojiContent:
                Text(c.emoji).font(.system(size: 34))
            case let c as FlareStickerContent:
                mediaThumb(c.url, side: 96, colors: colors, fallback: "square.on.square")
            case let c as FlareImageContent:
                imageBody(c.thumbnailURL ?? c.url, colors: colors)
            case let c as FlareVideoContent:
                videoBody(c, colors: colors)
            case let c as FlareAudioContent:
                Label(Self.duration(c.durationSec), systemImage: "speaker.wave.2")
                    .foregroundColor(onBubble)
            case let c as FlareFileContent:
                fileBody(c, fg: onBubble, colors: colors)
            case let c as FlareLocationContent:
                locationBody(c, fg: onBubble, colors: colors)
            case let c as FlareCardContent:
                cardBody(c, colors: colors)
            case let c as FlarePlaceholderContent:
                chip(c.label, colors: colors)
            case let c as FlareGenericContent:
                chip("[\(c.label)]", colors: colors)
            default:
                chip("[\(content.type)]", colors: colors)
            }
        }
    }

    private func imageBody(_ url: String, colors: FlareColors) -> some View {
        ZStack {
            asyncImage(url, colors: colors, fallback: "photo")
                .frame(maxWidth: 240, maxHeight: 240)
            if ctx.mediaState?.isDownloading == true {
                Color.black.opacity(0.35)
                Text("\(ctx.mediaState?.progressPct ?? 0)%").foregroundColor(.white).bold()
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: FlareSizes.radiusLg))
        .onTapGesture { ctx.onMediaAction?(content) }
    }

    private func videoBody(_ c: FlareVideoContent, colors: FlareColors) -> some View {
        ZStack {
            if let poster = c.poster, !poster.isEmpty {
                asyncImage(poster, colors: colors, fallback: "film").frame(maxWidth: 240, maxHeight: 240)
            } else {
                colors.bgTertiary.frame(width: 200, height: 130)
            }
            Image(systemName: "play.circle.fill").font(.system(size: 44)).foregroundColor(.white.opacity(0.9))
        }
        .clipShape(RoundedRectangle(cornerRadius: FlareSizes.radiusLg))
        .onTapGesture { ctx.onMediaAction?(content) }
    }

    private func fileBody(_ c: FlareFileContent, fg: Color, colors: FlareColors) -> some View {
        HStack {
            Image(systemName: "doc").font(.system(size: 26)).foregroundColor(fg)
            VStack(alignment: .leading) {
                Text(c.name).font(.system(size: FlareSizes.fontSizeLg, weight: .medium))
                    .foregroundColor(fg).lineLimit(1).frame(maxWidth: 180, alignment: .leading)
                if c.sizeBytes > 0 {
                    Text(Self.bytes(c.sizeBytes)).font(.system(size: FlareSizes.fontSizeSm))
                        .foregroundColor(fg.opacity(0.7))
                }
            }
        }
        .onTapGesture { ctx.onMediaAction?(content) }
    }

    private func locationBody(_ c: FlareLocationContent, fg: Color, colors: FlareColors) -> some View {
        HStack {
            Image(systemName: "mappin.circle").font(.system(size: 22)).foregroundColor(colors.error)
            VStack(alignment: .leading) {
                Text(c.name).font(.system(size: FlareSizes.fontSizeLg, weight: .medium)).foregroundColor(fg)
                if !c.address.isEmpty {
                    Text(c.address).font(.system(size: FlareSizes.fontSizeSm))
                        .foregroundColor(fg.opacity(0.7)).lineLimit(1).frame(maxWidth: 200, alignment: .leading)
                }
            }
        }
    }

    private func cardBody(_ c: FlareCardContent, colors: FlareColors) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            if let img = c.imageURL, !img.isEmpty {
                asyncImage(img, colors: colors, fallback: "photo").frame(height: 120).clipped()
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(c.title).font(.system(size: FlareSizes.fontSizeLg, weight: .semibold))
                    .foregroundColor(colors.textPrimary).lineLimit(2)
                if let sub = c.subtitle, !sub.isEmpty {
                    Text(sub).font(.system(size: FlareSizes.fontSizeSm))
                        .foregroundColor(colors.textSecondary).lineLimit(2)
                }
                if let src = c.sourceLabel, !src.isEmpty {
                    Text(src).font(.system(size: FlareSizes.fontSizeXs)).foregroundColor(colors.textTertiary)
                }
            }
            .padding(FlareSizes.spacingMd)
        }
        .frame(maxWidth: 240)
        .background(RoundedRectangle(cornerRadius: FlareSizes.radiusLg).fill(colors.bgPrimary))
        .overlay(RoundedRectangle(cornerRadius: FlareSizes.radiusLg).stroke(colors.borderSecondary))
    }

    private func chip(_ label: String, colors: FlareColors) -> some View {
        Text(label)
            .font(.system(size: FlareSizes.fontSizeSm)).foregroundColor(colors.textSecondary)
            .padding(.horizontal, FlareSizes.spacingSm).padding(.vertical, FlareSizes.spacingXs)
            .background(RoundedRectangle(cornerRadius: FlareSizes.radiusSm).fill(colors.bgTertiary))
    }

    private func mediaThumb(_ url: String, side: CGFloat, colors: FlareColors, fallback: String) -> some View {
        asyncImage(url, colors: colors, fallback: fallback).frame(width: side, height: side)
    }

    private func asyncImage(_ url: String, colors: FlareColors, fallback: String) -> some View {
        Group {
            if let u = URL(string: url), !url.isEmpty {
                AsyncImage(url: u) { image in
                    image.resizable().scaledToFill()
                } placeholder: {
                    fallbackTile(colors: colors, symbol: fallback)
                }
            } else {
                fallbackTile(colors: colors, symbol: fallback)
            }
        }
    }

    private func fallbackTile(colors: FlareColors, symbol: String) -> some View {
        ZStack {
            colors.bgTertiary
            Image(systemName: symbol).font(.system(size: 32)).foregroundColor(colors.textTertiary)
        }
    }

    static func duration(_ seconds: Int) -> String {
        String(format: "%02d:%02d", seconds / 60, seconds % 60)
    }

    static func bytes(_ b: Int) -> String {
        if b < 1024 { return "\(b) B" }
        if b < 1024 * 1024 { return String(format: "%.1f KB", Double(b) / 1024) }
        if b < 1024 * 1024 * 1024 { return String(format: "%.1f MB", Double(b) / 1024 / 1024) }
        return String(format: "%.1f GB", Double(b) / 1024 / 1024 / 1024)
    }
}

File: android-im-ui/src/main/kotlin/com/flare/im/ui/MessageContentView.kt
package com.flare.im.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.LocationOn
import androidx.compose.material.icons.outlined.Movie
import androidx.compose.material.icons.outlined.VolumeUp
import androidx.compose.material.icons.rounded.PlayCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Context passed to every content renderer. */
data class FlareContentContext(
    val isSelf: Boolean,
    val previewMode: Boolean = false,
    val senderName: String? = null,
    val mediaState: FlareMediaDownloadState? = null,
    val onMediaAction: ((FlareMessageContent) -> Unit)? = null,
)

typealias FlareContentBuilder = @Composable (FlareMessageContent, FlareContentContext) -> Unit

/**
 * Registry for product content types (`vote`, `task`…). Built-in types are
 * rendered directly by [MessageContentView]; register a builder to add/override.
 */
object FlareContentRegistry {
    private val builders = mutableMapOf<String, FlareContentBuilder>()
    fun register(type: String, builder: FlareContentBuilder) { builders[type] = builder }
    fun unregister(type: String) { builders.remove(type) }
    fun lookup(type: String): FlareContentBuilder? = builders[type]
}

/**
 * Content-type dispatcher — renders a message body by type. Spec:
 * Message/MessageContentView (`MessageContentView`). Image/video are shown as
 * placeholders (the package bundles no image loader); register a builder or wrap
 * with Coil in the host to display real media.
 */
@Composable
fun MessageContentView(
    content: FlareMessageContent,
    isSelf: Boolean = false,
    senderName: String? = null,
    mediaState: FlareMediaDownloadState? = null,
    onMediaAction: ((FlareMessageContent) -> Unit)? = null,
) {
    val ctx = FlareContentContext(isSelf, senderName = senderName, mediaState = mediaState, onMediaAction = onMediaAction)
    val custom = FlareContentRegistry.lookup(content.type)
    if (custom != null) { custom(content, ctx); return }

    val colors = flareColors()
    val onBubble = if (isSelf) Color.White else colors.textPrimary

    when (content) {
        is FlareTextContent -> Text(content.text, color = onBubble, fontSize = FlareSizes.fontSizeXl.value.sp, lineHeight = (FlareSizes.fontSizeXl.value * 1.45f).sp)
        is FlareEmojiContent -> FlareEmojiPackMessage(content.emoji, isSelf = ctx.isSelf)
        is FlareStickerContent -> FlareStickerPackMessage(
            stickerId = content.stickerId.orEmpty(),
            packageId = content.packageId,
            url = content.url,
            width = content.width,
            height = content.height,
            isSelf = ctx.isSelf,
        )
        is FlareImageContent -> Box {
            mediaPlaceholder(200.dp, colors, Icons.Outlined.Image, onClick = { onMediaAction?.invoke(content) })
            if (mediaState?.isDownloading == true) {
                Box(Modifier.size(200.dp).background(Color.Black.copy(alpha = 0.35f)), contentAlignment = Alignment.Center) {
                    Text("${mediaState.progressPct}%", color = Color.White, fontWeight = FontWeight.SemiBold)
                }
            }
        }
        is FlareVideoContent -> Box(contentAlignment = Alignment.Center) {
            mediaPlaceholder(200.dp, colors, Icons.Outlined.Movie, onClick = { onMediaAction?.invoke(content) })
            Icon(Icons.Rounded.PlayCircle, null, Modifier.size(44.dp), tint = Color.White.copy(alpha = 0.9f))
        }
        is FlareAudioContent -> Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Outlined.VolumeUp, null, tint = onBubble)
            Spacer(Modifier.width(FlareSizes.spacingSm))
            Text(duration(content.durationSec), color = onBubble, fontSize = FlareSizes.fontSizeLg.value.sp)
        }
        is FlareFileContent -> Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.clickable { onMediaAction?.invoke(content) },
        ) {
            Icon(Icons.Outlined.Description, null, Modifier.size(28.dp), tint = onBubble)
            Spacer(Modifier.width(FlareSizes.spacingSm))
            Column {
                Text(content.name, color = onBubble, fontWeight = FontWeight.Medium, maxLines = 1,
                    overflow = TextOverflow.Ellipsis, modifier = Modifier.widthIn(max = 180.dp))
                if (content.sizeBytes > 0) Text(bytes(content.sizeBytes), color = onBubble.copy(alpha = 0.7f), fontSize = FlareSizes.fontSizeSm.value.sp)
            }
        }
        is FlareLocationContent -> Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Outlined.LocationOn, null, tint = colors.error)
            Spacer(Modifier.width(FlareSizes.spacingXs))
            Column {
                Text(content.name, color = onBubble, fontWeight = FontWeight.Medium)
                if (content.address.isNotEmpty()) Text(content.address, color = onBubble.copy(alpha = 0.7f),
                    fontSize = FlareSizes.fontSizeSm.value.sp, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.widthIn(max = 200.dp))
            }
        }
        is FlareCardContent -> Column(
            Modifier.widthIn(max = 240.dp).clip(RoundedCornerShape(FlareSizes.radiusLg))
                .background(colors.bgPrimary).border(1.dp, colors.borderSecondary, RoundedCornerShape(FlareSizes.radiusLg))
                .padding(FlareSizes.spacingMd),
        ) {
            Text(content.title, color = colors.textPrimary, fontWeight = FontWeight.SemiBold, maxLines = 2, overflow = TextOverflow.Ellipsis)
            content.subtitle?.takeIf { it.isNotEmpty() }?.let {
                Text(it, color = colors.textSecondary, fontSize = FlareSizes.fontSizeSm.value.sp, maxLines = 2, overflow = TextOverflow.Ellipsis)
            }
            content.sourceLabel?.takeIf { it.isNotEmpty() }?.let {
                Spacer(Modifier.size(FlareSizes.spacingSm)); Text(it, color = colors.textTertiary, fontSize = FlareSizes.fontSizeXs.value.sp)
            }
        }
        is FlarePlaceholderContent -> chip(content.label, colors)
        is FlareGenericContent -> chip("[${content.label}]", colors)
        else -> chip("[${content.type}]", colors)
    }
}

@Composable
private fun chip(label: String, colors: FlareColors) {
    Box(
        Modifier.clip(RoundedCornerShape(FlareSizes.radiusSm)).background(colors.bgTertiary)
            .padding(horizontal = FlareSizes.spacingSm, vertical = FlareSizes.spacingXs),
    ) { Text(label, color = colors.textSecondary, fontSize = FlareSizes.fontSizeSm.value.sp) }
}

@Composable
private fun mediaPlaceholder(side: androidx.compose.ui.unit.Dp, colors: FlareColors, icon: androidx.compose.ui.graphics.vector.ImageVector, onClick: (() -> Unit)? = null) {
    Box(
        Modifier.size(side).heightIn(max = 200.dp).clip(RoundedCornerShape(FlareSizes.radiusLg)).background(colors.bgTertiary)
            .then(if (onClick != null) Modifier.clickable { onClick() } else Modifier),
        contentAlignment = Alignment.Center,
    ) { Icon(icon, null, Modifier.size(32.dp), tint = colors.textTertiary) }
}

internal fun duration(seconds: Int): String = "%02d:%02d".format(seconds / 60, seconds % 60)

internal fun bytes(b: Int): String = when {
    b < 1024 -> "$b B"
    b < 1024 * 1024 -> "%.1f KB".format(b / 1024.0)
    b < 1024 * 1024 * 1024 -> "%.1f MB".format(b / 1024.0 / 1024)
    else -> "%.1f GB".format(b / 1024.0 / 1024 / 1024)
}
